packages/flutter/lib/src/widgets/layout_builder.dart

```diff
@@ -256,42 +284,58 @@ class _LayoutBuilderElement<ConstraintType extends Constraints> extends RenderOb
 
   @override
   void removeRenderObjectChild(RenderObject child, Object? slot) {
-    final RenderConstrainedLayoutBuilder<ConstraintType, RenderObject> renderObject =
+    final RenderAbstractLayoutBuilderMixin<LayoutInfoType, RenderObject> renderObject =
         this.renderObject;
     assert(renderObject.child == child);
     renderObject.child = null;
     assert(renderObject == this.renderObject);
   }
 }
 
-/// Generic mixin for [RenderObject]s created by [ConstrainedLayoutBuilder].
+/// Generic mixin for [RenderObject]s created by an [AbstractLayoutBuilder] with
+/// the the same `LayoutInfoType`.
 ///
-/// Provides a callback that should be called at layout time, typically in
-/// [RenderObject.performLayout].
-mixin RenderConstrainedLayoutBuilder<
-  ConstraintType extends Constraints,
-  ChildType extends RenderObject
->
+/// Provides a [rebuildIfNecessary] method that should be called at layout time,
+/// typically in [RenderObject.performLayout]. The method invokes
+/// [AbstractLayoutBuilder]'s builder callback if needed.
+///
+/// Implementers must provide a [layoutInfo] implementation that is safe to
+/// access in [rebuildIfNecessary], which is typically called in [performLayout].
+mixin RenderAbstractLayoutBuilderMixin<LayoutInfoType, ChildType extends RenderObject>
     on RenderObjectWithChildMixin<ChildType> {
-  LayoutCallback<ConstraintType>? _callback;
+  LayoutCallback<Constraints>? _callback;
 
   /// Change the layout callback.
-  void updateCallback(LayoutCallback<ConstraintType>? value) {
+  void _updateCallback(LayoutCallback<Constraints>? value) {
     if (value == _callback) {
       return;
     }
     _callback = value;
     markNeedsLayout();
   }
 
-  /// Invoke the callback supplied via [updateCallback].
+  /// Invoke the builder callback supplied via [AbstractLayoutBuilder] and
+  /// rebuilds the [AbstractLayoutBuilder]'s widget tree, if needed.
+  ///
+  /// No work will be done if [layoutInfo] has not changed since the last time
+  /// this method was called, and [AbstractLayoutBuilder.updateShouldRebuild]
+  /// returned `false` when the widget was rebuilt.
   ///
-  /// Typically this results in [ConstrainedLayoutBuilder.builder] being called
-  /// during layout.
+  /// This method should typically be called as soon as possible in the class's
+  /// [performLayout] implementation, before any layout work is done.
+  @protected
   void rebuildIfNecessary() {
     assert(_callback != null);
     invokeLayoutCallback(_callback!);
   }
+
+  /// The (subset of) information to invoke the `builder` callback with.
```

Comment:
nit: lets specify here where one can find the builder callback since it isn't on this class (e.g. link this to [AbstractLayoutBuilder.builder]).